### PR TITLE
[SPARK-9103][WIP] Add Memory Tracking UI and track Netty memory usage

### PR DESCRIPTION
**Pull request by jsoltren**
This patch resurrects https://github.com/apache/spark/pull/7753 by liyezhang556520.

## What changes were proposed in this pull request?

Memory reporting is a much desired feature in Apache Spark. This change adds reporting for some memory used by netty, and adds a new UI tab for memory reporting. We introduce a new executorMetrics object to track this.

## How was this patch tested?

Ran https://github.com/jsoltren/spark-demos/blob/master/src/main/scala/org/soltren/spark/examples/ShuffleGroupCount.scala and verified appearance of new tab in UI.

9103-memory-ui

### CODE REVIEW

File: core/src/main/scala/org/apache/spark/network/netty/NettyBlockTransferService.scala [hunk 1]
```diff
@@ -22,8 +22,12 @@ import java.nio.ByteBuffer
 import scala.collection.JavaConverters._
 import scala.concurrent.{Future, Promise}
 import scala.reflect.ClassTag
+import scala.tools.nsc.interpreter.JList
```
Comment (squito) [inline core/src/main/scala/org/apache/spark/network/netty/NettyBlockTransferService.scala]: this should be

```scala
import java.util.{List => JList}
```

(it happens to work b/c of this: https://github.com/scala/scala/blob/v2.12.1/src/repl/scala/tools/nsc/interpreter/package.scala#L36)

File: core/src/main/scala/org/apache/spark/scheduler/EventLoggingListener.scala [hunk 2]
```diff
@@ -87,6 +88,10 @@ private[spark] class EventLoggingListener(
   // Visible for tests only.
   private[scheduler] val logPath = getLogPath(logBaseDir, appId, appAttemptId, compressionCodecName)
 
+  private val executorIdToLatestMetrics = new HashMap[String, SparkListenerExecutorMetricsUpdate]
+  private val executorIdToModifiedMaxMetrics = new
+    HashMap[String, SparkListenerExecutorMetricsUpdate]
```
Comment (squito) [inline core/src/main/scala/org/apache/spark/scheduler/EventLoggingListener.scala]: nit: put `new` on same line as `hashMap`

File: core/src/main/scala/org/apache/spark/scheduler/EventLoggingListener.scala [hunk 3]
```diff
@@ -259,6 +290,46 @@ private[spark] class EventLoggingListener(
...
+   * @param latestEvent the latest event received, used to update our map of stored metrics.
+   */
+  private def updateModifiedMetrics(latestEvent: SparkListenerExecutorMetricsUpdate): Unit = {
+    val executorId = latestEvent.execId
+    val toBeModifiedEvent = executorIdToModifiedMaxMetrics.get(executorId)
+    toBeModifiedEvent match {
+      case None =>
+        executorIdToModifiedMaxMetrics(executorId) = latestEvent
+      case Some(toBeModifiedEvent) =>
+        val toBeModifiedTransportMetrics = toBeModifiedEvent.executorMetrics.transportMetrics
```
Comment (squito) [inline core/src/main/scala/org/apache/spark/scheduler/EventLoggingListener.scala]: how about `prevTransportMetrics`

File: core/src/main/scala/org/apache/spark/scheduler/EventLoggingListener.scala [hunk 4]
```diff
@@ -259,6 +290,46 @@ private[spark] class EventLoggingListener(
...
+          toBeModifiedTransportMetrics.onHeapSize
+        }
+        val offHeapSize =
+            if (latestTransportMetrics.offHeapSize > toBeModifiedTransportMetrics.offHeapSize) {
+          timeStamp = latestTransportMetrics.timeStamp
+          latestTransportMetrics.offHeapSize
+        } else {
+          toBeModifiedTransportMetrics.offHeapSize
+        }
+        val modifiedExecMetrics = ExecutorMetrics(toBeModifiedEvent.executorMetrics.hostname,
```
Comment (squito) [inline core/src/main/scala/org/apache/spark/scheduler/EventLoggingListener.scala]: how about `updatedExecMetrics`

**Comment by squito** [comment]
Jenkins, ok to test

**Comment by jerryshao** [comment]
@jsoltren thanks to bring up this very old PR.

By looking at the UI you pasted here, I'm wondering what is the usage of `Completed Stages` here, what's difference here compared to `Stages` tab?

Also I think it would be better to to add a new Metrics Source about Netty memory usage, that would be useful.

**Comment by tgravescs** [comment]
we also just exposed more memory information for storage memory in the executors page in SPARK-17019. 

If we now have a memory tab it could be confusing to the users where to go to see what. If this is again a per executor thing do we just add new columns which check boxes there or consider consolidating that information under this memory tab?

File: core/src/test/scala/org/apache/spark/util/JsonProtocolSuite.scala [hunk 5]
```diff
@@ -432,6 +439,25 @@ class JsonProtocolSuite extends SparkFunSuite {
...
+        .accumulators().map(AccumulatorSuite.makeInfo)
+        .zipWithIndex.map { case (a, i) => a.copy(id = i) }
+    val executorMetricsUpdate = SparkListenerExecutorMetricsUpdate("exec3", new ExecutorMetrics,
+      Seq((1L, 2, 3, accumUpdates)))
+    assert(executorMetricsUpdate.executorMetrics != null)
+    assert(executorMetricsUpdate.executorMetrics.transportMetrics != null)
+    val newJson = JsonProtocol.executorMetricsUpdateToJson(executorMetricsUpdate)
+    val oldJson = newJson.removeField { case (field, _) => field == "Executor Metrics Updated"}
+    val newMetrics = JsonProtocol.executorMetricsUpdateFromJson(oldJson)
+    assert(newMetrics.executorMetrics.hostname === "")
```
Comment (squito) [inline core/src/test/scala/org/apache/spark/util/JsonProtocolSuite.scala]: it seems like really `executorMetrics` should be an `Option` (with a comment that it was not present till version x)

Comment (squito) [inline core/src/test/scala/org/apache/spark/util/JsonProtocolSuite.scala]: it seems like really `executorMetrics` should be an `Option` (with a comment that it was not present till version x)

**Comment by squito** [comment]
yeah taking another look at the UI, I agree with @jerryshao and @tgravescs, the memory tab is pretty weird.  I think putting this info on the "stages" tab makes sense -- that is really the info that this is currently exposing, the memory used by netty during a stage.  I think at some point it might make sense to have a Memory tab which more completely summarizes the memory used by spark -- netty, cache, shuffle, execution, perhaps even parquet & serialization buffers, etc.  But with just, its more confusing than it is helpful.

It is also a little bit strange having just this one bit of memory info on the stage page until we get a more complete picture of memory, but hopefully still useful.

Also having a metric source for netty seems like a good idea.

**Comment by squito** [comment]
@tgravescs what do you think about breaking this into two parts -- the internal plumbing, and the UI stuff?  by itself the plumbing part wouldn't do anything, but I think it would be easier to review, as long as there is general agreement this is a good direction.

**Comment by tgravescs** [comment]
I haven't looked through the code at all, but I definitely like the idea of tracking the netty memory usage.  Breaking into 2 pieces makes sense.

If we end up creating any new UI pages please use datatables like the executors page  rather then the old format.

**Comment by jerryshao** [comment]
@jsoltren , a quick look at your current implementation, looks like you only track the Netty memory usage in `NettyBlockTransferService`, but in Spark there're some other places which will create Netty clientFactory or server, I think it would be better to also track memory usage in those places:

1. Netty RPC client factory and RPC Server.
2. Netty file download client factory.
3. Netty external shuffle client factory.
4. Netty block transfer client factory and server - currently you did this already. For the server I think it is not used for external shuffle.

Also Spark has Netty context (rpc, shuffle), shall we also need different netty metrics for shuffle, rpc...?

I would suggest to only expose Netty metrics internally in this PR and get rid of UI things for the following PR.

**Comment by jsoltren** [comment]
I'm still making updates here. It would be ideal if one of the admins could please re-open this PR. Otherwise I will open a new one. Thanks!

**Comment by jsoltren** [comment]
This PR was closed, so, I'll create a new one focusing on just the back end pieces. I'll create a fresh JIRA for more general memory tracking improvements to the UI where we can hash out more of the details. The UI has changed quite a lot since the original PR!

**Pull request closed without merging**